fastlane.swift: Write your Fastfile in swift

taquitos opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Many iOS developers want to use fastlane, but are resistant to try it because it is written in ruby. Ruby doesn’t work with Xcode, so people find it more cumbersome to edit and test. Asking our developers to understand a new language, a new framework, and no autocomplete, creates a barrier to entry. Most engineers of a statically typed language have a common pattern for learning new languages/frameworks: they start typing and rely on the IDE to auto-complete. This isn’t available to new users of fastlane.
- What if we could meet the developer where they are: Xcode?
- What if they could test their fastfile syntax before they tried to run the commands?
- What if you had syntax highlighting and autocompletion?
- What if you could use the same language that you use to build apps, for fastlane?
- What if you could do all this while also enabling other languages to be used?

### Description
✅ fastlane gains an embedded socket server
✅ Swift implementation of a socket client
✅ fastlane generates Swift code that utilizes the socket client during deployment of fastlane
✅ Swift code organized into a CLI app that references a user’s files (Fastfile.swift, Appfile.swift, etc) and compiles them into FastlaneRunner
✅ FastlaneRunner is executed from fastlane if a Fastfile.swift is detected
✅ FastlaneRunner is rebuilt if Fastfile.swift (or any other file like Appfile.swift) is has a last modified date after FastlaneRunner was built.
✅ fastlane automatically rebuilds and tests swift runner before deployment

Commit:
Added basics for calling actions from generated swift

Commit:
Added return types to all actions with clear return types
Modified Swift generator to account for return types

Commit:
Create subclass structures for Lanefile subclasses

Commit:
Added Default Swift Fastfile
Improved type inference from ruby to Swift types
Updated handling of environment variables in Swift

Commit:
Autogenerate all the various tools protocols and default implementations
Ignore some default param values that are derived from the environment
Added processing of Options types

Commit:
Exclude precheck rules from Swift api since they aren't actually options yet
Generate parsing functions for action return values
Pass back objects from executed commands and use generated parsing functions
Added logger with verbose/normal options for Swift Runner

Commit:
Added calling of beforeAll/afterall
Add ability to run fastlane socket server like: $fastlane socket_server

Commit:
Updated swift project to include AppFile
added run script phase to copy FastlaneRunner after building to source dir

Commit:
Updated deliver to include Deliver.swift
Updated appFile to include AppFile.swift
Piped is_swift throughout setup
can now run fastlane init swift to init for Fastlane.swift

Commit:
Updated Precheck to include Precheckfile.swift
can call fastlane precheck init swift now

Commit:
Updated Gym to include Gymfile.swift
Can now call fastlane gym init swift

Commit:
Updated Match to include Matchfile.swift
can now call fastlane match init swift

Commit:
Updated scan to include Scanfile.swift
can now call fastlane scan init swift

Commit:
Updated screengrab to include Screengrabfile.swift
can now call fastlane screengrab init swift

Commit:
Updated snapshot to include Snapfile.swift
can now run fastlane snapshot init swift

Commit:
Added swift lane manager
Added gym option to disable archiving
FastlaneRunner now automatically built
Launching of Fastlane.swift now automatic

Commit:
Default implementations of each <Toolname>file.swift

Commit:
Detect which <Tool>file need to be replaced in the xcodeproj
Replace default <Tool>file when found
Move default <Tool>file back when user removes their own

Commit:
Add support for error_callback callbacks
Add return type of :string to sh.rb
JSON escape return objects so multi-line strings work properly

Commit:
Added special handling for 'sh' action to match the fast_file.rb
Updated fast_file.rb so self.sh can be reused
Added method run_action_requiring_special_handling for actions that will
require special handling due to overrides in fast_file.rb that can't be
inferred from the action source code

Commit:
Integrated building and testing of generated Swift into release process
Updated FastlaneRunner project to include a default Fastfile.swift and Appfile.swift

Organized FastlaneRunner project so it's easier to understand what is going on
Updated Swift templates to strip out Ruby mentions
Improved fatalError() messages for runner since fastlane hides log output
Added `timeoutSeconds` arg to runner in case 1 hour isn't enough time
Added messaging to `fastlane init swift` so you know where to edit the Fastfile.swift

Squash me

Commit:
Implemented ./FastlaneRunner lanes command

Commit:
Added support for command: fastlane lanes (for swift)
Added time stamps and version stamps for autogenerated code
Updated Fastfile to delete the FastlaneRunner binary after building

Add detection of swift fastlane lanes and descriptions
Clean up FastfileTemplate

Commit:
Added example of desc() usage in the fastfile template
Removed old way of using desc()

Commit:
Updated API generation to add @discardableResult to all methods that return data

Commit:
Update template

Commit:
Remove _ = since we're now adding @discardableResult
Clean up some messages in deliverfiledefault

Commit:
Added environment variable setting and getting

Commit:
Improve analytics so we can differentiate between ruby and swift fastfiles
Add list of actions to not include in fastlane.swift
update String.characters usage to String
Remove Environment Variable stuff from Swift, rely on ruby action

ohayon committed:
handle string array arguments (#11054)
replace print statements with UI.verbose

Commit:
Centralize type inference

ohayon committed:
separate out params to new lines (#11080)
* separate out params to new lines
* add comment
* give the arguments each their own line too

Commit:
Add type: Hash to ConfigItems missing it

ohayon committed:
allow handling of [String: Any] to be valid json for passing to ruby (#11093)

* allow handling of [String: Any] to be valid json for passing to ruby

Commit:
Update to support versioning individual files

Commit:
Add support for Hash type in ConfigItem
Update all generated files with better documentation
Update to remove emoji usage from reserved words

ohayon committed:
Use JSONSerialization for encoding the string before sending it to the socket server (#11118)

Commit:
Created Fastlane::Boolean so we can generate smarter Swift API

ohayon committed:
add quotes to run script phase for spaces (#11126)
* add quotes to run script phase for spaces

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

taquitos commented on deliver/lib/assets/DeliverfileDefault:
```diff
@@ -1,10 +1,10 @@
 ###################### More Options ######################
 # If you want to have even more control, check out the documentation
-# https://docs.fastlane.tools/actions/deliver
```
Ugh, merge issue. Fixing 🌮

janpio commented on deliver/lib/assets/DeliverfileDefault:
```diff
@@ -1,10 +1,10 @@
 ###################### More Options ######################
 # If you want to have even more control, check out the documentation
-# https://docs.fastlane.tools/actions/deliver
+# https://github.com/fastlane/fastlane/blob/master/deliver/Deliverfile.md
```
Do you really want to set this back to Github?

janpio commented on deliver/lib/assets/DeliverfileDefault.swift:
```diff
@@ -0,0 +1,13 @@
+/************************ More Options ************************/
+// If you want to have even more control, check out the documentation
+// https://github.com/fastlane/fastlane/blob/master/deliver/Deliverfile.md
```
Github? Not docs.fastlane.tools?

revolter replied:
Replace with `https://docs.fastlane.tools/actions/deliver`.

janpio commented on fastlane/lib/assets/AppfileTemplate.swift:
```diff
@@ -0,0 +1,8 @@
+var itcTeam: String? { return [[ITC_TEAM]] } // iTunes Connect Team ID (may be nil if no team)
+
+// you can even provide different app identifiers, Apple IDs and team names per lane:
+// More information: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md
```
correct link?

revolter replied:
Replace with `https://docs.fastlane.tools/advanced/#appfile`.

janpio commented on fastlane/lib/assets/DefaultFastfileTemplate.swift:
```diff
@@ -0,0 +1,73 @@
+// Customize this file, documentation can be found here:
+// https://github.com/fastlane/fastlane/tree/master/fastlane/docs
```
correct link?

revolter commented on fastlane/lib/assets/DefaultFastfileTemplate.swift:
```diff
@@ -0,0 +1,73 @@
+        // )
+    }
+
+    // More information about multiple platforms in fastlane: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Platforms.md
```
Replace with `https://docs.fastlane.tools/advanced/#control-configuration-by-lane-and-by-platform`.

taquitos commented:
@revolter I think I updated all the doc links 🤦‍♂️

janpio commented on fastlane/lib/fastlane/actions/cocoapods.rb:
```diff
@@ -100,13 +100,14 @@ def self.available_options
                                        description: 'Retry with --repo-update if action was finished with error',
                                        optional: true,
                                        is_string: false,
-                                       default_value: true,
```
wasn't necessary?

taquitos replied:
it was causing a conflict with `repo_update` since that is set to `false` as default.

milch replied:
Shouldn't it be the other way around, i.e. only update the repo when there is an error?

taquitos replied:
Yeah, but I don't think it's plumbed that way right now, so we'd have to update how the relatively newish `try_repo_update_on_error` interacts with the `repo_update` option.

janpio commented on fastlane/lib/fastlane/actions/environment_variable.rb:
```diff
@@ -0,0 +1,73 @@
+module Fastlane
```
new action? why?

taquitos replied:
Great question, if you want to set environment variable in Fastfile.swift, the only way to get _fastlane_ to actually use them, is to pass them along. That's because they operate as 2 separate processes and so the ruby process doesn't inherit the environment.

janpio replied:
Something for that technical blog post ;)

mpirri replied:
Can we add that justification either to a comment or the description of this action? Just so that folks know when to use this vs using ENV[].

janpio commented on fastlane/lib/fastlane/actions/sh.rb:
```diff
@@ -1,3 +1,5 @@
+
+require_relative '../helper/sh_helper.rb'
```
required, but not used?

taquitos replied:
That might be some cruft left over from when I was working on special-casing the `sh` action. I'll take a look at that.

janpio commented on gym/lib/gym/options.rb:
```diff
@@ -129,6 +133,13 @@ def self.plain_options
                                      is_string: false,
+                                     type: Boolean,
+                                     optional: true),
+        FastlaneCore::ConfigItem.new(key: :skip_archive,
```
new feature?

taquitos replied:
Yeah, we're using gym to build the `FastlaneRunner` binary after Swift API generation, but gym really wants to create a fully ITC capable package.

Commit:
Add configurable 5 second default connection timeout
Add option to keep server alive after error/disconnect (debug usage)
Pass --verbose through to FastlaneRunner (logMode verbose)
Pass username and app_id through onboarding

ohayon commented on fastlane/lib/fastlane/commands_generator.rb:
```diff
@@ -151,11 +151,26 @@ def run
       command :socket_server do |c|
         c.syntax = 'fastlane start_server'
         c.description = 'Starts local socket server and enables only a single local connection'
+        c.option '-s', '--stay_alive', 'Keeps socket server up even after error or disconnects, requires CTRL-C to kill.'
```
🕺

ohayon replied:
get it?

revolter replied:
> Feel the city breakin' and everybody shakin' 🎶

taquitos replied:
![huehuehue](https://user-images.githubusercontent.com/1304821/33776584-b72cc964-dbf6-11e7-8ca5-7e0aab7b4483.gif)

KrauseFx commented on fastlane/lib/fastlane/lane_manager_base.rb:
```diff
@@ -0,0 +1,122 @@
+module Fastlane
+  class LaneManagerBase
```
It would be 💯 to have a 1-2 sentence comment here on what class is responsible for, and what to use it for

KrauseFx commented on fastlane/lib/fastlane/server/command.rb:
```diff
@@ -0,0 +1,72 @@
+      end
+
+      # rubocop:disable Security/Eval
+      return eval(class_name)
```
Loading a class with `eval` seems a little off. Couldn't we use the following 
```ruby
Fastlane::Actions.const_get(class_name)
```
^ in this case we'd load a class that's inside the module `Fastlane::Actions`

KrauseFx commented on fastlane/lib/fastlane/server/command_executor.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+  end
+
+  class GenericCommandExecuter < CommandExecutor
```
Comments on what the class does would be 💯, also usually we avoid having 2 classes in one file, can we add a comment on why we needed to do it?

taquitos replied:
This class is going to be removed 🌮 updated the class that is staying with comments

ohayon reviewed (approved):
This PR is 💯 💥 💣 👍 🚀 

Just left a few comments, mostly about some names and extracting out conditional switches on whether or not the config is written in swift into their own methods. Nothing pressing to block a merge from me tho!

mpirri commented on fastlane/lib/fastlane/server/socket_server.rb:
```diff
@@ -0,0 +1,212 @@
+    end
+  end
+
+  class JSONReturnValueProcessor
```
Any reason for this class to not get its own file?

mpirri commented:
This looks 💯! 🥇
